design/baremetal-operator/bmo-part-of-capm3.md

```diff
@@ -0,0 +1,283 @@
+├── rbac
+│   ├── kustomization.yaml
+│   ├── role_binding.yaml
+│   ├── role.yaml
+│   └── service_account.yaml
+└── role.yaml -> rbac/role.yaml
+
+```
+
+As we can see, the `deploy` directory has one top level folder for deployment,
+namely `default` and it deploys only baremetal-operator through kustomization
+file and uses kustomization config file for teaching kustomize where to look at
+when substituting variables. In addition, `crds`, `namespace` and `rbac` folders
+have their own kustomization and yaml files. The following directory tree
+visualizes a new structure of ironic-deployment folder in more detail:
+
+## Proposed structure of ironic-deployment folder
+
+```diff
+tree ironic-deployment/
+
+ironic-deployment/
+├── default
+│   ├── ironic_bmo_configmap.env
+│   └── kustomization.yaml
+├── ironic
+│   ├── ironic.yaml
+│   └── kustomization.yaml
+└── keepalived
+    ├── ironic_bmo_configmap.env
+    ├── keepalived_patch.yaml
+    └── kustomization.yaml
+
+```
+
+In the above provided tree, ironic-deployment folder has three top level folders
+for deployments, namely `default`, `ironic` and `keepalived`. `default` and
+`ironic` folders will deploy only ironic, whereas, `keepalived` folder deploys
+the ironic with keepalived. As the name implies, `keepalived/keepalived_patch.yaml`
+patches the default image URL through kustomization. The user should run the
+following commands to be able to meet requirements of each use case as provided
+below:
+
+### Commands to deploy "production" use case after structural changes
+
+```diff
+kustomize build $BMOPATH/deploy/default | kubectl apply -f-
+kustomize build $BMOPATH/ironic-deployment/default | kubectl apply -f-
+
+```
+
+### Command to deploy "local developer" use case after structural changes
+
+```diff
+kustomize build $BMOPATH/ironic-deployment/default | kubectl apply -f-
+
+```
+
+### Command to deploy "ironic outside of the cluster" use case after structural changes
+
+```diff
+kustomize build $BMOPATH/deploy/default | kubectl apply -f-
+
+```
+
+where $BMOPATH points to the baremetal-operator path.
+
+To this end, "production", "local developer" and "ironic outside of the cluster"
+use cases will be met, as new structure will fulfill the needs of those cases
+with cleaner and leaner organization of deployments in place.
+
+## Pivoting use case
+
+Once above outlined folder structural changes are made, we can deploy BMO on its
+own in CAPM3 and Ironic independently to meet the use case of pivoting, where
+BMO should be deployed as part of CAPM3. The main concern here is that, BMO and
+Ironic are separated and as a consequence, it will result in breaking the chain
+of workflow, where, previously, BMO could use localhost to talk to ironic
+without any authentication. However, considering that Ironic runs in host
+network, BMO would be able to interact with it, from anywhere that is connected
+or routed to the provisioning network. Therefore, separating Ironic in its own
```

Comment:
Today we run the BMO in the pod with Ironic and dnsmasq using host networking. After this change, we probably won't need that, since only dnsmasq and Ironic need access to the host network. That means the BMO will need to be able to see Ironic via the cluster's external network, which means linking the external and provisioning networks via a route or having Ironic listening on both networks.

Reply:
Maybe we can specify that this requires the host on which the BMO container is running to have access to the provisioning network ?